common/web/gesture-recognizer/src/cumulativePathStats.ts

```diff
@@ -0,0 +1,702 @@
+        }
+      }
+
+      result.sampleCount -= subsetStats.sampleCount;
+
+      // NOTE: baseSample MUST REMAIN THE SAME.  All math is based on the corresponding diff.
+      // Though... very long touchpoint interactions could start being affected by that "catastrophic
+      // cancellation" effect without further adjustment.  (If it matters, we'll get to that later.)
+      // But _probably_ not; we don't go far beyond a couple of orders of magnitude from the origin in
+      // ANY case except the timestamp (.t) - and even then, not far from the baseSample's timestamp value.
+
+      // initialSample, though, we need to update b/c of the 'directness' properties.
+      result.initialSample = subsetStats.followingSample;
+
+      return result;
+    }
+
+    public get lastSample() {
+      return this._lastSample;
+    }
+
+    public get lastTimestamp(): number {
+      return this.lastSample?.t;
+    }
+
+    public get sampleCount() {
+      return this._sampleCount;
+    }
+
+    private set sampleCount(value: number) {
+      this._sampleCount = value;
+    }
+
+    /**
+     * In order to mitigate the accumulation of small floating-point errors during the
+     * various accumulations performed by this class, the domain of incoming values
+     * is remapped near to the origin via axis-specific mapping constants.
+     * @param dim
+     * @returns
+     */
+    private mappingConstant(dim: 'x' | 'y' | 't' | 'v') {
+      if(!this.baseSample) {
+        return undefined;
+      }
+
+      if(dim == 't') {
+        return this.baseSample.t;
+      } else if(dim == 'x') {
+        return this.baseSample.targetX;
+      } else if(dim == 'y') {
+        return this.baseSample.targetY;
+      } else {
+        return 0;
+      }
+    }
+
+    /**
+     * Gets the statistical mean, utilizing the internal 'mapped' coordinate space.
+     * This is the version compatible with cross-sums and squared-sums.
+     * @param dim
+     * @returns
+     */
+    private mappedMean(dim: 'x' | 'y' | 't' |'v') {
+      return this.rawLinearSums[dim] / this.sampleCount;
+    }
+
+    /**
+     * Gets the statistical mean value of the samples observed during the represented
+     * interval on the specified axis.
+     * @param dim
+     * @returns
+     */
+    public mean(dim: 'x' | 'y' | 't' | 'v') {
+      // This external-facing version needs to provide values in 'external'-friendly
+      // coordinate space.
+      return this.mappedMean(dim) + this.mappingConstant(dim);
+    }
+
+    /**
+     * Gets the sum of the squared distance from the mean seen in samples observed
+     * during the represented interval on the specified axis.
+     * @param dim
+     * @returns
+     */
+    public squaredSum(dim: 'x' | 'y' | 't' | 'v') {
+      const x2 = this.rawSquaredSums[dim];
+      const x1 = this.rawLinearSums[dim];
+
+      const val = x2 - x1 * x1 / this.sampleCount;
+
+      return val > 1e-8 ? val : 0;
```

Comment:
After a bit more thought... this is happening as the squared-sums values increase in magnitude.

`Number.EPSILON` (see link above) is based on the integer `1`; what would be an equivalent value for sums at 1e6 & (1e6+1)?  Well... approximately `Number.EPSILON * 1e6`.

Fortunately, screen resolution hasn't gone off the deep end yet; taking 4000x3000, that gives us 1.2e7.  Supposing that gets summed up 100 times (due to a long hold or whatever), then we start needing to worry about increasing the floating-point error threshold.

------

So, if we want to be super-careful... I could instead implement an adaptive epsilon that considers the order of magnitude of the numbers being diffed?

Reply:
- As floating-point precision's cutoff is ~ 2e-52, perhaps an adaptive threshold like ...

```javascript
const crossSumDiff = operand1 + operand 2;
// If the difference is 2^30 smaller (> 1e9 smaller) than one of the original operands...
if(Math.log2(operand1) - Math.log2(crossSumDiff) > 30) {
    return 0;
} else {
    return crossSumDiff;
```

That'd probably work.  Granted, `Math.log2` is noted as a not-cheap (though still O(1)) operation.

Referencing https://scicomp.stackexchange.com/a/37322, it's not like we're constantly calling it... so maybe we can get away with using it?  It _is_ used as input to a branch... but at least it's not likely to "thrash" branch prediction efforts.

Reply:
I don't know the answer to these questions!

Reply:
I'll go ahead with that proposed code for now.  It'll still give us better precision than 32-bit floats.  (` > 23` instead of `> 30` would be about equivalent to 32-bit floats.)